annotate.el

```diff
@@ -1565,19 +1572,32 @@ annotation."
         (ignore-errors (%load-annotation-data))
       (%load-annotation-data))))
 
-(defun annotate-dump-annotation-data (data)
+(defun annotate-dump-annotation-data (data &optional save-empty-db)
   "Save `data` into annotation file."
-  (with-temp-file annotate-file
-    (let* ((print-length nil)
-           (%abbreviate-filename (lambda (record)
-                                   (let ((full-filename (annotate-filename-from-dump    record))
-                                         (annotations   (annotate-annotations-from-dump record))
-                                         (file-checksum (annotate-checksum-from-dump    record)))
-                                     (annotate-make-record (abbreviate-file-name full-filename)
-                                                           annotations
-                                                           file-checksum))))
-           (actual-data (mapcar %abbreviate-filename data)))
-      (prin1 actual-data (current-buffer)))))
+  (if (or save-empty-db
+          data)
+      (with-temp-file annotate-file
+        (let* ((print-length nil)
+               (%abbreviate-filename (lambda (record)
+                                       (let ((full-filename (annotate-filename-from-dump    record))
+                                             (annotations   (annotate-annotations-from-dump record))
+                                             (file-checksum (annotate-checksum-from-dump    record)))
+                                         (annotate-make-record (abbreviate-file-name full-filename)
+                                                               annotations
+                                                               file-checksum))))
+               (actual-data (mapcar %abbreviate-filename data)))
+          (prin1 actual-data (current-buffer))))
+    (let* ((confirm-message    "Delete annotations database file %S? [y/N] ")
+           (delete-confirmed-p (or (not annotate-database-confirm-deletion)
+                                   (string= (read-from-minibuffer (format confirm-message
+                                                                           annotate-file))
+                                            "y"))))
+      (if delete-confirmed-p
+          (condition-case err
+              (delete-file annotate-file t)
+            (error (message "error removing annotation database: %S"
+                            (error-message-string err))))
+        (annotate-dump-annotation-data data t)))))
```

Comment:
Why the additional recursion here (and the new optional argument)? I'm curious: did you run into a case where you were able to write data to the file, but not delete it?

Reply:
Hi!

The  recursive  call  is  made  when  the  user  totally  deleted  the
annotations in all  their files but asked not to  delete the database file,
in this case an empty (contains only  'nil') db is saved, this is the old
behaviour actually.

Or at least, this are the intentions of the code i wrote ;-) but seems
to work in my testing.

Bye!
C.

Reply:
Hi!

The  recursive  call  is  made  when  the  user  totally  deleted  the
annotations in all  their files but asked not to  delete the database file,
in this case an empty (contains only  'nil') db is saved, this is the old
behaviour actually.

Or at least, this are the intentions of the code i wrote ;-) but seems
to work in my testing.

Bye!
C.